### PR TITLE
clean up unused dependency repo's work trees

### PR DESCRIPTION
**Pull request by OsmondJiang**
@yufeih , how can we update the travis ci git version? the git worktree remove need 2.17xx git

### CODE REVIEW

File: docs/specs/restore.yml [hunk 1]
```diff
@@ -3,9 +3,25 @@
 inputs:
   docfx.yml: |
     dependencies:
-      docfx-test-dependencies: https://github.com/docascode/docfx-test-dependencies
+      docfx-test-dependencies1: https://github.com/docascode/docfx-test-dependencies
+      docfx-test-dependencies2: https://github.com/docascode/docfx-test-dependencies
 restores:
   git/github.com/docascode/docfx-test-dependencies/9ab39126f6d6642131084e741ca6206989db505a/dep.md: DEP
   git/github.com/docascode/docfx-test-dependencies/5736374c0a142252063bbedafe5d98db495e263d/dep.md: child-DEP
+  restore_lock/01. Restore dependencies with their children-*-lock.json: |
```
Comment (yufeih) [inline docs/specs/restore.yml]: >lock [](start = 62, length = 4)

I suppose this is a single file containing all transitive dependencies, like package-lock.json ?

File: test/docfx.Test/e2e/E2ETest.cs [hunk 2]
```diff
@@ -37,13 +37,16 @@ public static class E2ETest
...
+            if (spec.Restore)
+                await Program.Run(new[] { "restore", docsetPath });
+
             await Program.Run(new[] { "build", docsetPath });
 
             // Verify restored files
             foreach (var (file, content) in spec.Restores)
             {
-                var restoredFile = Path.Combine(AppData.AppDataDir, file);
+                var restoredFile = Directory.EnumerateFiles(AppData.AppDataDir, file, SearchOption.TopDirectoryOnly).FirstOrDefault();
```
Comment (yufeih) [inline test/docfx.Test/e2e/E2ETest.cs]: >FirstOrDefault [](start = 117, length = 14)

Single() ?

File: docs/specs/restore.yml [hunk 3]
```diff
@@ -3,9 +3,25 @@
...
+    {"git":{"https://github.com/docascode/docfx-test-dependencies":"9ab39126f6d6642131084e741ca6206989db505a"}}
+  restore_lock/9ab39126f6d6642131084e741ca6206989db505a-*-lock.json: |
+    {"git":{"https://github.com/docascode/docfx-test-dependencies#child":"5736374c0a142252063bbedafe5d98db495e263d"}}
 outputs:
   build.manifest:
+---
+# Restored dependency repo can't be found
+restore: false
```
Comment (superyyrrzz) [inline docs/specs/restore.yml]: what about use "commands: build", to indicate we don't need restore. "restore" and "restores" are a bit confusing.

File: src/docfx/restore/Restore.cs [hunk 4]
```diff
@@ -125,19 +133,72 @@ private static async Task<(string workTreePath, string workTreeHead)> FetchOrClo
...
-                    var workTrees = await GitUtility.ListWorkTrees(restorePath);
-                    if (!workTrees.Contains(workTreePath))
+                    var existingWorkTrees = new HashSet<string>(await GitUtility.ListWorkTrees(restorePath, false));
+                    var newWorkTrees = new ConcurrentBag<string>();
+                    await ParallelUtility.ForEach(hrefs, async href =>
+                    {
+                        var (_, rev) = GetGitRemoteInfo(href);
+                        var workTreeHead = await GitUtility.Revision(restorePath, rev);
+                        var workTreePath = GetRestoreWorkTreeDir(restoreDir, workTreeHead);
+                        if (existingWorkTrees.Add(workTreePath))
```
Comment (superyyrrzz) [inline src/docfx/restore/Restore.cs]: >Add [](start = 46, length = 3)

looks not thread safety?

File: docs/specs/restore.yml (hunk 1 above)
Comment (OsmondJiang) [reply docs/specs/restore.yml]: currently no requirement to put all transitive dependencies in one docset lock file, so let's just do that in this way and change it later once we got a requirement for that.

---
In reply to: [201566709](https://github.com/dotnet/docfx/pull/3042#discussion_r201566709) [](ancestors = 201566709)

File: docs/specs/restore.yml (hunk 3 above)
Comment (OsmondJiang) [reply docs/specs/restore.yml]: good suggestion

---
In reply to: [201568836](https://github.com/dotnet/docfx/pull/3042#discussion_r201568836) [](ancestors = 201568836)

File: src/docfx/restore/Restore.cs [hunk 5]
```diff
@@ -3,14 +3,18 @@
...
 using System.Diagnostics;
 using System.IO;
+using System.Linq;
 using System.Threading.Tasks;
 
 namespace Microsoft.Docs.Build
 {
     internal static class Restore
     {
+        private const int MaxWorkTreeCount = 5;
```
Comment (yufeih) [inline src/docfx/restore/Restore.cs]: >MaxWorkTreeCount [](start = 26, length = 16)

make this configurable using environment variable?

File: src/docfx/restore/Restore.cs [hunk 6]
```diff
@@ -91,53 +95,121 @@ public static string GetRestoreRootDir(string url)
...
                 });
 
-            Debug.Assert(!string.IsNullOrEmpty(workTreeHead));
-            Debug.Assert(!string.IsNullOrEmpty(workTreePath));
+                return (existingWorkTrees.Keys.ToList(), newWorkTrees.ToList());
+            }
+
+            bool NeedCleanupWorkTrees(int existingWorkTreeCount) => existingWorkTreeCount > MaxWorkTreeCount;
+
+            async Task CleanupWorkTrees(List<string> existingWorkTrees, List<string> newWorkTrees)
```
Comment (yufeih) [inline src/docfx/restore/Restore.cs]: >CleanupWorkTrees [](start = 23, length = 16)

Log a progress when clean up work trees.

File: src/docfx/restore/Restore.cs (hunk 5 above)
Comment (OsmondJiang) [reply src/docfx/restore/Restore.cs]: I don't think the end user need to know it, let's just hardcode it and make it configurable when we want to implement `docfx gc` function?

---
In reply to: [201593922](https://github.com/dotnet/docfx/pull/3042#discussion_r201593922) [](ancestors = 201593922)

File: test/docfx.Test/restore/RestoreTest.cs [hunk 7]
```diff
@@ -27,5 +28,43 @@ public static void GetGitInfo(string remote, string expectedUrl, string expected
...
+dependencies:
+  dep1: {gitUrl}#test-clean-1
+  dep2: {gitUrl}#test-clean-2
+  dep3: {gitUrl}#test-clean-3
+  dep4: {gitUrl}#test-clean-4
+  dep5: {gitUrl}#master
+  dep6: {gitUrl}#chi");
+
+            // run restroe and check the work trees
+            await Restore.Run(docsetPath, new CommandLineOptions(), new Report());
```
Comment (yufeih) [inline test/docfx.Test/restore/RestoreTest.cs]: >Restore [](start = 18, length = 7)

Program.Run("restore") ?

File: src/docfx/restore/Restore.cs (hunk 5 above)
Comment (yufeih) [reply src/docfx/restore/Restore.cs]: ok

---
In reply to: [201595705](https://github.com/dotnet/docfx/pull/3042#discussion_r201595705) [](ancestors = 201595705,201593922)

**Pull request merged**